docs/plugin-hierarchy.md

```diff
@@ -0,0 +1,28 @@
+## Plugin hierarchy
+
```

Comment:
Sure we can! I'm not sure if we can call any of our plugins "internal" because they introduce tasks and stuff that clients depend on. All plugin classes are internal but the behavior they introduce (like tasks they add) is public and we need to keep it compatible.